apollo-federation/src/schema/schema_upgrader.rs

```diff
@@ -25,308 +23,160 @@ use crate::error::FederationError;
+                            TypeInfo {
+                                pos: pos.clone(),
+                                metadata: subgraph.metadata().clone(), // TODO: Prefer not to clone
+                            },
+                        );
+                }
             }
         }
-    }
 
-    if !subgraphs_using_interface_object.is_empty() {
-        // TODO: Make this a composition error and make the strings "human readable"
-        let cond_1_str = subgraphs_using_interface_object
+        let subgraphs: HashMap<String, Subgraph<Expanded>> = subgraphs
             .iter()
-            .cloned()
-            .map(|k| format!("\"{k}\""))
-            .collect::<Vec<_>>()
-            .join(" ");
-        let cond_2_str = upgraded
-            .keys()
-            .map(|k| format!("\"{k}\""))
-            .collect::<Vec<_>>()
-            .join(" ");
-        return Err(internal_error!(
-            "The @interfaceObject directive can only be used if all subgraphs have federation 2 subgraph schema (schema with a `@link` to \"https://specs.apollo.dev/federation\" version 2.0 or newer): @interfaceObject is used in subgraph {} but subgraph {} is not a federation 2 subgraph schema.",
-            cond_1_str,
-            cond_2_str
-        ));
-    }
-    Ok(subgraphs
-        .into_iter()
-        .map(|(name, subgraph)| {
-            if !subgraph.metadata().is_fed_2_schema() {
-                return upgraded.remove(&name).unwrap();
-            }
-            subgraph
-        })
-        .collect())
-}
-
-// Extensions for FederationError to provide additional error types
-impl FederationError {
-    pub fn extension_with_no_base(message: &str) -> Self {
-        // Fixed: use internal() instead of new()
-        FederationError::internal(format!("EXTENSION_WITH_NO_BASE: {}", message))
-    }
-}
-
-impl<'a> SchemaUpgrader<'a> {
-    #[allow(unused)]
-    fn new(
-        original_subgraph: &'a Subgraph<Expanded>,
-        subgraphs: &'a IndexMap<String, Subgraph<Expanded>>,
-        object_type_map: &'a HashMap<Name, HashMap<String, TypeInfo>>,
-    ) -> Result<Self, FederationError> {
-        let schema = original_subgraph.schema().clone();
-        // TODO (FED-428): JS version calls `setSchemaAsFed2Subgraph`.
-        Ok(SchemaUpgrader {
-            schema,
-            expanded_info: ExpandedSubgraphInfo {
-                subgraph_name: original_subgraph.name.clone(),
-                subgraph_url: original_subgraph.url.clone(),
-                key_directive_name: original_subgraph.key_directive_name()?.clone(),
-                requires_directive_name: original_subgraph.requires_directive_name()?.clone(),
-                provides_directive_name: original_subgraph.provides_directive_name()?.clone(),
-                extends_directive_name: original_subgraph.extends_directive_name()?.clone(),
-            },
+            .map(|subgraph| (subgraph.name.clone(), subgraph.clone()))
```

Comment:
On the whole, I like the reorganization, but the downside is that we're introducing an extra clone for all schemas here(even those that are not fed1) which the old code was avoiding.

Reply:
Indeed but it also keeps the code "cleaner". If this turns out a performance problem we can always optimize later.